File: make_movie.m
function [] = make_movie(filename)
%MAKE_MOVIE Make a movie from bim_test data
load('alpha.mat', 'alpha')

window_x=1.2;
window_y=1.2;

fig = figure(); set(gca,'FontSize',18); set(gcf,'color','w');

%% -------- save plot to movie --------
VW = VideoWriter([filename '.avi']);
VW.FrameRate = 30;
open(VW);

for i = 1:99
    fname = 'test_' + string(i) + '.mat';
    load(fname, 'positions', 'thetas', 'U_n');
    t = 100*0.001*i;
    
    %% update theta and L
    normals = [-sin(thetas); cos(thetas)];
    area_n = 3.186673437718255;
    
    %% conserved quantities
    area_np2 = 0.5*trapzp(positions(1,:).^2+positions(2,:).^2, length(alpha)); % integral of r dr dtheta

    %% plot & save image
    clf;
    plot([positions(1,:) positions(1,1)],[positions(2,:) positions(2,1)],'bo-','LineWidth',2); grid on;
    axis([-window_x window_x -window_y window_y]); axis equal; hold on;
    plot(sqrt(area_n/pi)*[cos(alpha) cos(alpha(1))],sqrt(area_n/pi)*[sin(alpha) sin(alpha(1))],'r-');
    %plot(cm_np2(1),cm_np2(2),'r.','MarkerSize',10);
    title([' t = ' num2str(t) ' , ' 'A/A_0 = ' num2str(area_np2/area_n)]);
    q2 = quiver(positions(1,:),positions(2,:),U_n.*normals(1,:),U_n.*normals(2,:)); q2.AutoScale = 'off';
    frame = getframe(fig);
    writeVideo(VW,frame);
end
close(fig)

end

File: make_movie.py
#!/usr/bin/env python3

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import h5py
from matplotlib.animation import FFMpegWriter

with h5py.File("test.h5", "r") as f:
    positions_t = np.array(f["positions_t"])
    theta_t = np.array(f["theta_t"])
    U_t = np.array(f["U_t"])
    alpha = np.array(f["alpha"])
    area_n = np.array(f["area_n"])[0]
    n_record = np.array(f["n_record"])[0]
    dt = np.array(f["dt"])[0]

fig = plt.figure()
metadata = dict(title="Movie Test", artist="Me", comment="Testing. Duh.")
writer = FFMpegWriter(fps=30, metadata=metadata)

with writer.saving(fig, "test.mp4", positions_t.shape[0]):
    for i in range(0, positions_t.shape[0]):
        positions = positions_t[i, :, :]
        thetas = theta_t[i, :]
        U_n = U_t[i, :]

        normals = np.vstack([-np.sin(thetas), np.cos(thetas)]).T

        plt.plot(
            np.sqrt(area_n / np.pi) * np.hstack([np.cos(alpha), np.cos(alpha[0])]),
            np.sqrt(area_n / np.pi) * np.hstack([np.sin(alpha), np.sin(alpha[0])]),
            color="red",
        )
        plt.plot(
            np.hstack([positions[:, 0], positions[:, 0]]),
            np.hstack([positions[:, 1], positions[:, 1]]),
            color="blue",
        )
        plt.quiver(
            positions[:, 0],
            positions[:, 1],
            U_n * normals[:, 0],
            U_n * normals[:, 1],
            scale=2.5,
            color="blue",
            headwidth=1.5,
            headlength=2.5,
        )
        plt.xlim([-1.2, 1.2])
        plt.ylim([-1.2, 1.2])
        writer.grab_frame()

        plt.cla()
